lib/hal_client/link.rb

```diff
@@ -0,0 +1,137 @@
+require 'hal_client/representation'
+
+class HalClient
+
+  # HAL representation of a single link. Provides access to an embedded representation.
+  class Link
+
+    # Create a new Link
+    #
+    # options - name parameters
+    #   :rel - This Link's rel property
+    #   :target - An instance of Representation
+    #   :template - A URI template ( https://www.rfc-editor.org/rfc/rfc6570.txt )
+    #   :curie_resolver - An instance of CurieResolver (used to resolve curied rels)
+    def initialize(options)
+      @literal_rel = options[:rel]
+      @target = options[:target]
+      @template = options[:template]
+      @curie_resolver = options[:curie_resolver] || CurieResolver.new([])
+
+      (fail ArgumentError, "A rel must be provided") if @literal_rel.nil?
+
+      if @target.nil? && @template.nil?
+        (fail ArgumentError, "A target or template must be provided")
+      end
+
+      if @target && @template
+        (fail ArgumentError, "Cannot provide both a target and a template")
+      end
+
+      if @target && !@target.kind_of?(Representation)
+        (fail ArgumentError, "Invalid HAL representation: #{target.inspect}")
+      end
+
+      if @template && !@template.kind_of?(Addressable::Template)
+        (fail ArgumentError, "Invalid Addressable::Template: #{template.inspect}")
+      end
+    end
+
+    attr_accessor :literal_rel, :target, :template, :curie_resolver
+
+
+    # Create a new Link using an entry from the '_links' section of a HAL document
+    #
+    # options - name parameters
+    #   :hash_entry - a hash containing keys :rel (string) and :data (hash from a '_links' entry)
+    #   :hal_client - an instance of HalClient
+    #   :curie_resolver - An instance of CurieResolver (used to resolve curied rels)
+    #   :base_url - Base url for resolving relative links in hash_entry (probably the parent
+    # document's "self" link)
+    def self.new_from_link_entry(options)
+      hash_entry = options[:hash_entry]
+      hal_client = options[:hal_client]
+      curie_resolver = options[:curie_resolver]
+      base_url = options[:base_url]
+
+      rel = hash_entry[:rel]
+      hash_data = hash_entry[:data]
+      href = (Addressable::URI.parse(base_url) + hash_data['href']).to_s
```

Comment:
does this work for anonymous resources? ie, resources without a self link or explicit href? Maybe we should pass in an already parsed url and move the defaulting behavior (https://github.com/charlieok/hal-client/blob/21b833dd5efabe0d138f8217f41bec10ddb5c57e/lib/hal_client/representation.rb#L338) to a `#base_url` method on representation.